File: cpp/Source/gamit/serialize/encrypt.cpp
#include "encrypt.h"
#include <random>

using namespace gamit;

byte_t CEncrypto::_mask = 108;

void CEncrypto::simpleEncrypt(std::string & src)
{
	__mask(src);
	__encrypt(src);
}

void CEncrypto::simpleDecrypt(std::string & src)
{
	__decrypt(src);
	__mask(src);
}

void CEncrypto::__mask(std::string & src)
{
	if (src.empty())
	{
		return;
	}

	unsigned maxIdx = src.size() - 1;

	for (unsigned i = 0; i <= maxIdx; i += 2)
	{
		if (i == maxIdx)
		{
			src[i] ^= _mask;
			return;
		}

		byte_t bi = src[i];
		byte_t bj = src[i + 1];

		bi ^= _mask;
		bj ^= _mask;

		src[i] = bj;
		src[i + 1] = bi;
	}
}

void CEncrypto::__encrypt(std::string & src)
{
	if (src.empty())
	{
		return;
	}

	byte_t pivot = byte_t(std::rand() % 127 + 1);

	for (byte_t i = 0; i < src.size(); ++i)
	{
		src[i] ^= pivot;
	}

	src.push_back(pivot);
}

void CEncrypto::__decrypt(std::string & src)
{
	if (src.empty())
	{
		return;
	}

	byte_t pivot = *src.rbegin();
	src.pop_back();

	for (std::string::size_type i = 0; i < src.size(); ++i)
	{
		src[i] ^= pivot;
	}
}

File: python/gamit/serialize/encrypt.py
"""
* @name encrypt.py
*
* @author ahda86@gmail.com
*
* @date 2015/6/25 19:17
*
* @desc encrypt.py
"""

import random

def __mask(src):
    mask = 108

    buffSize = len(src)
    if buffSize == 0:
        return

    maxIdx = buffSize - 1
    for i in range(0, buffSize, 2):
        if i == maxIdx:
            src[i] ^= mask
            return

        bi = src[i]
        bj = src[i + 1]

        bi ^= mask
        bj ^= mask

        src[i] = bj
        src[i + 1] = bi

def __encrypt(src):
    buffSize = len(src)

    if buffSize == 0:
        return

    pivot = random.randint(1, 127)
    src.append(pivot)

    for i in range(buffSize - 1, -1, -1):
        src[i] ^= pivot
        pivot = src[i]

def __decrypt(src):
    buffSize = len(src)

    if buffSize == 0:
        return

    for i in range(0, buffSize - 1):
        src[i] ^= src[i + 1]

    del src[-1]

def simpleEncrypt(src):
    dest = list(src)
    __mask(dest)
    __encrypt(dest)

    return bytes(dest)

def simpleDecrypt(src):
    dest = list(src)

    __decrypt(dest)
    __mask(dest)

    return bytes(dest)

